File: 0121. Best Time to Buy and Sell Stock/Solution0121.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution0121 {
public:
    int maxProfit(vector<int>& prices) {
        int maxProf = 0;
        int n = prices.size();
        int minCost = INT_MAX;
        for (int i = 0; i < n; i++) {
            if (prices[i] <= minCost) {
                minCost = prices[i];
            } else {
                maxProf = max(maxProf, prices[i] - minCost);
            }
        }
        return maxProf;
    }
};

File: 0121. Best Time to Buy and Sell Stock/Solution0121.py
from typing import List


class Solution0121:
    def maxProfit(self, prices: List[int]) -> int:
        res, curmin = 0, prices[0]
        n = len(prices)
        for i in range(n):
            res = max(res, prices[i] - curmin)
            curmin = min(curmin, prices[i])
        return res